src/surrogate_heat_driver.cpp

```diff
@@ -87,24 +164,148 @@ void SurrogateHeatDriver::generate_arrays()
+  // be steady-state or pseudo-steady-state cases with no axial conduction such that
+  // the power deposition in each channel is independent of a convective heat transfer
+  // coefficient and only depends on the rod power at that axial elevation. The channel
+  // powers are indexed by channel ID, axial ID
+  xt::xtensor<double, 2> channel_powers = xt::zeros<double>({n_channels_, n_axial_});
+  for (int i = 0; i < n_channels_; ++i) {
+    for (int j = 0; j < n_axial_; ++j) {
+      for (const auto& rod : channels_[i].rod_ids_)
+        channel_powers(i, j) += 0.25 * rod_axial_node_power(rod, j);
+    }
   }
 
+  // initial guesses for the fluid solution are uniform temperature (set to the inlet
+  // temperature) and uniform pressure  (set to the outlet pressure). These solution
+  // fields are defined on channel axial faces
+  xt::xtensor<double, 2> h = xt::zeros<double>({n_channels_, n_axial_ + 1});
+  xt::xtensor<double, 2> p = xt::zeros<double>({n_channels_, n_axial_ + 1});
+  std::fill(h.begin(), h.end(), iapws::h1(pressure_bc_, inlet_temperature_));
+  std::fill(p.begin(), p.end(), pressure_bc_);
+
+  bool converged = false;
+  for (int iter = 0; iter < max_subchannel_its_; ++iter) {
+    // save the previous solution
+    xt::xtensor<double, 2> h_old = h;
+    xt::xtensor<double, 2> p_old = p;
+
+    // solve each channel independently
+    for (int chan = 0; chan < n_channels_; ++chan) {
+      const auto& c = channels_[chan];
+
+      // solve for enthalpy by simple energy balance q = mdot * dh by marching from inlet
+      h(chan, 0) = iapws::h1(p(chan, 0), inlet_temperature_);
+      for (int axial = 0; axial < n_axial_; ++axial)
+        h(chan, axial + 1) = h(chan, axial) + channel_powers(chan, axial) / channel_flowrates_(chan);
+
+      // solve for pressure using one-sided finite difference approximation by marching
+      // from outlet and solving the axial momentum equation.
+      p(chan, n_axial_) = pressure_bc_;
+      for (int axial = n_axial_; axial > 0; axial--) {
+        double rho_high = 1.0e-3 / iapws::nu1(p[axial], iapws::T_from_p_h(p[axial], h[axial]));
+        double rho_low = 1.0e-3 / iapws::nu1(p[axial - 1], iapws::T_from_p_h(p[axial - 1], h[axial - 1]));
+
+        double u_high = channel_flowrates_(chan) / (rho_high * c.area_);
+        double u_low = channel_flowrates_(chan) / (rho_low * c.area_);
+
+        p[axial - 1] = p[axial] + channel_flowrates_(chan) / c.area_ * (u_high - u_low) +
+          g_ * (z_(axial) - z_(axial - 1)) * rho_low;
+      }
+    }
+
+    // after solving all channels, check for convergence; although all channels are independent,
+    // to enable crossflow coupling between channels in the future, this convergence check is
+    // performed on all channels together, rather than each separately, since in a more sophisticated
+    // solver the channels would all be linked
+    auto n_expr_h = xt::norm_l1(h - h_old);
+    double h_norm = n_expr_h();
```

Comment:
Since `n_expr_h` is not used again, I would just combine these two lines (same below as well)